### PR TITLE
Support FUNCTION [...] commands

### LINKED ISSUE #2876: Support new FUNCTION [...] commands
- [x] [FCALL](https://redis.io/commands/fcall)
- [x] [FCALL_RO](https://redis.io/commands/fcall_ro)
- [x] [FUNCTION DELETE](https://redis.io/commands/function-delete)
- [x] [FUNCTION DUMP](https://redis.io/commands/function-dump)
- [x] [FUNCTION FLUSH](https://redis.io/commands/function-flush)
- [x] [FUNCTION KILL](https://redis.io/commands/function-kill)
- [x] [FUNCTION LIST](https://redis.io/commands/function-list)
- [x] [FUNCTION LOAD](https://redis.io/commands/function-load)
- [x] [FUNCTION RESTORE](https://redis.io/commands/function-restore)
- [ ] [FUNCTION STATS](https://redis.io/commands/function-stats)

### LINKED ISSUE #2893: Add support for redis command: FCALL

### LINKED ISSUE #2894: Add support for redis command: FCALL_RO

### LINKED ISSUE #2895: Add support for redis command: FUNCTION DELETE (R7)

### LINKED ISSUE #2896: Add support for redis command: FUNCTION DUMP (R7)

### LINKED ISSUE #2897: Add support for redis command: FUNCTION FLUSH (R7)

### LINKED ISSUE #2898: Add support for redis command: FUNCTION KILL (R7)

### LINKED ISSUE #2899: Add support for redis command: FUNCTION LIST (R7)

### LINKED ISSUE #2900: Add support for redis command: FUNCTION LOAD (R7)

### LINKED ISSUE #2901: Add support for redis command: FUNCTION RESTORE (R7)

### LINKED ISSUE #2902: Add support for redis command: FUNCTION STATS (R7)

### PR DESCRIPTION
**Pull request by Avital-Fine**
Closes #2876, Closes #2893, Closes #2894, Closes #2895, Closes #2896, Closes #2897, Closes #2898, Closes #2899, Closes #2900, Closes #2901 Closes #2902

### CODE REVIEW

File: src/main/java/redis/clients/jedis/resps/LibraryInfo.java [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+
+    public String getEngine() {
+        return engine;
+    }
+
+    public String getDescription() {
+        return description;
+    }
+
+    public FunctionInfo[] getFunctions() {
```
Comment (sazzad16) [inline src/main/java/redis/clients/jedis/resps/LibraryInfo.java:62]: DON'T use array.

File: src/main/java/redis/clients/jedis/params/FunctionLoadParams.java [hunk 2]
```diff
@@ -0,0 +1,40 @@
+package redis.clients.jedis.params;
+
+import static redis.clients.jedis.Protocol.Keyword.REPLACE;
+import static redis.clients.jedis.Protocol.Keyword.DESC;
+import redis.clients.jedis.CommandArguments;
```
Comment (sazzad16) [inline src/main/java/redis/clients/jedis/params/FunctionLoadParams.java:5]:
```suggestion
import static redis.clients.jedis.Protocol.Keyword.REPLACE;
import static redis.clients.jedis.Protocol.Keyword.DESC;

import redis.clients.jedis.CommandArguments;
```

**Review by sazzad16 — CHANGES_REQUESTED** [review]
Also, in all new files, you've used 4-spaces as tabs. Please use 2-spaces as tabs.

**Comment by Avital-Fine** [comment]
1. Can anyone help me to test FUNCTION KILL? The client is waiting to `fcall` replay so I can't execute `functionKill`...
2. Another problem is to test `fcallReadonly`: L434 in the test throws `ERR Can not execute a script with write flag using *_ro command.`

File: src/main/java/redis/clients/jedis/commands/FunctionBinaryCommands.java [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+
+import redis.clients.jedis.args.FunctionRestorePolicy;
+import redis.clients.jedis.params.FunctionLoadParams;
+import redis.clients.jedis.resps.LibraryInfo;
+
+import java.util.List;
+
+public interface FunctionBinaryCommands {
+
+  byte[] fcall(byte[] name, List<byte[]> keys, List<byte[]> args);
```
Comment (zeekling) [inline src/main/java/redis/clients/jedis/commands/FunctionBinaryCommands.java:11]:
```suggestion
/**
   * Invoke a function.
   * @param name
   * @param keys
   * @param args
   * @return
   */
  byte[] fcall(byte[] name, List<byte[]> keys, List<byte[]> args);
```

File: src/test/java/redis/clients/jedis/commands/jedis/ScriptingCommandsTest.java [hunk 4]
```diff
@@ -315,6 +319,133 @@ public void scriptExistsWithBrokenConnection() {
...
+
+  @Test
+  public void functionKill() {
+    jedis.functionFlush();
+    String engine = "Lua";
+    String library = "mylib";
+    String function = "redis.register_function('myfunc', function(keys, args)\n local a = 1 while true do a = a + 1 end \nend)";
+
+    jedis.functionLoad(engine, library, function);
+//    jedis.fcall("myfunc", new ArrayList<>(), new ArrayList<>());
```
Comment (zeekling) [inline src/test/java/redis/clients/jedis/commands/jedis/ScriptingCommandsTest.java:425]: del unused code

**Review by zeekling — APPROVED** [review]

File: src/test/java/redis/clients/jedis/commands/jedis/ScriptingCommandsTest.java (hunk 4 above)
Comment (sazzad16) [reply src/test/java/redis/clients/jedis/commands/jedis/ScriptingCommandsTest.java:425]: @zeekling She wants to keep this test and asked for help to make this test work...

> Can anyone help me to test FUNCTION KILL? The client is waiting to `fcall` reply so I can't execute `functionKill`...

Comment (zeekling) [reply src/test/java/redis/clients/jedis/commands/jedis/ScriptingCommandsTest.java:449]: get it

**Comment by sazzad16** [comment]
@yangbodong22011 @dengliming Please review.

File: src/test/java/redis/clients/jedis/commands/jedis/AllKindOfValuesCommandsTest.java [hunk 5]
```diff
@@ -671,7 +671,7 @@ public void pexpire() {
     assertEquals(1, jedis.pexpire("foo2", 200000000000L));
     assertEquals(0, jedis.pexpire("foo2", 10000000, ExpiryOption.NX));
     assertEquals(1, jedis.pexpire("foo2", 10000000, ExpiryOption.XX));
-    assertEquals(0, jedis.pexpire("foo2", 10000000, ExpiryOption.GT));
+    assertEquals(0, jedis.pexpire("foo2", 10000, ExpiryOption.GT));
```
Comment (sazzad16) [inline src/test/java/redis/clients/jedis/commands/jedis/AllKindOfValuesCommandsTest.java:674]: Is it necessary? If so, do in separate PR.

File: src/test/java/redis/clients/jedis/commands/unified/AllKindOfValuesCommandsTestBase.java [hunk 6]
```diff
@@ -466,7 +466,7 @@ public void pexpire() {
     assertEquals(1, jedis.pexpire("foo2", 200000000000L));
     assertEquals(0, jedis.pexpire("foo2", 10000000, ExpiryOption.NX));
     assertEquals(1, jedis.pexpire("foo2", 10000000, ExpiryOption.XX));
-    assertEquals(0, jedis.pexpire("foo2", 10000000, ExpiryOption.GT));
+    assertEquals(0, jedis.pexpire("foo2", 10000, ExpiryOption.GT));
```
Comment (sazzad16) [inline src/test/java/redis/clients/jedis/commands/unified/AllKindOfValuesCommandsTestBase.java:469]: Is it necessary? If so, do in separate PR.

**Review by sazzad16 — CHANGES_REQUESTED** [review]
@Avital-Fine

**Review by sazzad16 — APPROVED** [review]

File: src/main/java/redis/clients/jedis/resps/LibraryInfo.java [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+
+  public List<Map<String, Object>> getFunctions() {
+    return functions;
+  }
+
+  public String getLibraryCode() {
+    return libraryCode;
+  }
+
+  public static final Builder<LibraryInfo> LIBRARY = new Builder<LibraryInfo>() {
```
Comment (Avital-Fine) [inline src/main/java/redis/clients/jedis/resps/LibraryInfo.java:52]: Maybe `LIBRARY_BUILDER` will be better?

File: src/main/java/redis/clients/jedis/commands/FunctionBinaryCommands.java [hunk 8]
```diff
@@ -0,0 +1,131 @@
...
+   * that did not modify the dataset during their execution.
+   * @return OK
+   */
+  String functionKill();
+
+  /**
+   * Return information about the functions and libraries.
+   * @return {@link LibraryInfo}
+   */
+  List<Object> functionListBinary();
```
Comment (Avital-Fine) [inline src/main/java/redis/clients/jedis/commands/FunctionBinaryCommands.java:64]: Can we move this function to `FunctionCommands.java`?

File: src/main/java/redis/clients/jedis/commands/FunctionBinaryCommands.java [hunk 9]
```diff
@@ -0,0 +1,131 @@
...
+
+  /**
+   * Return information about the functions and libraries.
+   * @param libraryNamePattern a pattern for matching library names
+   * @return {@link LibraryInfo}
+   */
+  List<Object> functionList(byte[] libraryNamePattern);
```
Comment (Avital-Fine) [inline src/main/java/redis/clients/jedis/commands/FunctionBinaryCommands.java:71]: Can we move this function to `FunctionCommands.java`?

File: src/main/java/redis/clients/jedis/commands/FunctionBinaryCommands.java [hunk 10]
```diff
@@ -0,0 +1,131 @@
...
+
+  /**
+   * Similar to {@link FunctionCommands#functionList() FUNCTION LIST} but include the
+   * libraries source implementation in the reply.
+   * @see FunctionCommands#functionList()
+   * @return {@link LibraryInfo}
+   */
+  List<Object> functionListWithCodeBinary();
```
Comment (Avital-Fine) [inline src/main/java/redis/clients/jedis/commands/FunctionBinaryCommands.java:79]: Can we move this function to `FunctionCommands.java`?

File: src/test/java/redis/clients/jedis/commands/jedis/ScriptingCommandsTest.java [hunk 11]
```diff
@@ -315,6 +321,187 @@ public void scriptExistsWithBrokenConnection() {
...
+  @Test
+  public void functionLoadAndDelete() {
+    String engine = "Lua";
+    String library = "mylib";
+    String function = "redis.register_function('myfunc', function(keys, args) return args[1] end)";
+
+    assertEquals("OK", jedis.functionLoad(engine, library, function));
+    assertEquals("OK", jedis.functionLoad(engine, library, new FunctionLoadParams().replace().libraryDescription(""), function));
+
+    jedis.functionDelete(library);
```
Comment (Avital-Fine) [inline src/test/java/redis/clients/jedis/commands/jedis/ScriptingCommandsTest.java:333]:
```suggestion
    assertEquals("OK", jedis.functionDelete(library));
```

Comment (Avital-Fine) [inline src/test/java/redis/clients/jedis/commands/jedis/ScriptingCommandsTest.java:333]:
```suggestion
    assertEquals("OK", jedis.functionDelete(library));
```

File: src/test/java/redis/clients/jedis/commands/jedis/ScriptingCommandsTest.java [hunk 12]
```diff
@@ -315,6 +321,187 @@ public void scriptExistsWithBrokenConnection() {
...
+
+    // Binary
+    assertEquals("OK", jedis.functionLoad(engine.getBytes(), library.getBytes(), function.getBytes()));
+    assertEquals("OK", jedis.functionLoad(engine.getBytes(), library.getBytes(), new FunctionLoadParams().replace(), function.getBytes()));
+
+    jedis.functionDelete(library.getBytes());
```
Comment (Avital-Fine) [inline src/test/java/redis/clients/jedis/commands/jedis/ScriptingCommandsTest.java:339]:
```suggestion
    assertEquals("OK", jedis.functionDelete(library.getBytes()));
```

File: src/test/java/redis/clients/jedis/commands/jedis/ScriptingCommandsTest.java [hunk 13]
```diff
@@ -315,6 +321,187 @@ public void scriptExistsWithBrokenConnection() {
...
+
+    // check with code and with LIBRARYNAME
+    response = jedis.functionListWithCode(library).get(0);
+    assertEquals(library, response.getLibraryName());
+    assertEquals(function, response.getLibraryCode());
+
+    // Binary
+    List<Object> bresponse = (List<Object>) jedis.functionListBinary().get(0);
+    assertArrayEquals(library.getBytes(), (byte[]) bresponse.get(1));
+
```
Comment (Avital-Fine) [inline src/test/java/redis/clients/jedis/commands/jedis/ScriptingCommandsTest.java:392]:
```suggestion
    List<Object> bresponse = (List<Object>) jedis.functionListWithCodeBinary().get(0);
    assertArrayEquals(library.getBytes(), (byte[]) bresponse.get(1));
```
```
